### PR TITLE
Datalog: Optimize factsOf some more

### PR DESCRIPTION
**Pull request by bdahse**
Remove another `Array.map` call.

### CODE REVIEW

File: main/src/library/Fixpoint/Solver.flix [hunk 1]
```diff
@@ -45,8 +45,8 @@ namespace Fixpoint {
          lowerStmt |>
          interpret |>
          toAst) as & Pure;
-         let _ = notifyPostSolve(model);
-         model
```
Comment (magnus-madsen) [inline main/src/library/Fixpoint/Solver.flix:49]: Why the change in indentation?

File: main/src/library/Fixpoint/Solver.flix [hunk 2]
```diff
@@ -399,44 +399,44 @@ namespace Fixpoint {
...
+        let f = terms -> (unbox(headTermValue(terms[0])), unbox(headTermValue(terms[1])), unbox(headTermValue(terms[2])), unbox(headTermValue(terms[3])), unbox(headTermValue(terms[4])));
         factsOf(f, p, d) as & Pure
 
     ///
     /// Returns an array of facts associated with the given predicate symbol `p` in the given Datalog program `d`.
     ///
-    def factsOf(f: Array[v] -> t & e, p: PredSym, d: Datalog[v]): Array[t] & Impure =
+    def factsOf(f: Array[HeadTerm[v]] -> t & e, p: PredSym, d: Datalog[v]): Array[t] & Impure =
         let Datalog(facts, _) = d;
         restrict(facts, Set#{p}) |>
```
Comment (magnus-madsen) [inline main/src/library/Fixpoint/Solver.flix:438]: Presumably restrict traverses all facts once and filters them. Perhaps `filterMap` can be used. Or alternatively, for even more speed, you could use `Array.foreach` and a `MutList` to accumulate the result?

File: main/src/library/Fixpoint/Solver.flix (hunk 1 above)
Comment (bdahse) [reply main/src/library/Fixpoint/Solver.flix:49]: Brings the let binding in line with the other let bindings.

File: main/src/library/Fixpoint/Solver.flix (hunk 2 above)
Comment (bdahse) [reply main/src/library/Fixpoint/Solver.flix:438]: Yes, both are possible. However `Array.filterMap` has not been optimized like `Array.filter`. Why do you expect `MutList` to be faster than a (optimized) `filterMap`?

Comment (magnus-madsen) [reply main/src/library/Fixpoint/Solver.flix:438]: Only because it avoids lots of allocation of `Some` objects.

Comment (bdahse) [reply main/src/library/Fixpoint/Solver.flix:438]: Fair point. Would `MutList.reserve!` make sense to have an array that's large enough upfront or would that be too space inefficient? It would have same length as `facts`.
